src/main/cpp/profiler/ProfilerJni.cpp

```diff
@@ -0,0 +1,530 @@
+    // State->has_cupti_callback_errored = true;
+    std::cerr << "PROFILER: Error handling callback: " << get_cupti_error(rc) << std::endl;
+    return;
+  }
+
+  auto cb_data = static_cast<CUpti_CallbackData const*>(callback_data_ptr);
+  switch (domain) {
+    case CUPTI_CB_DOMAIN_DRIVER_API: domain_driver_callback(callback_id, cb_data); break;
+    case CUPTI_CB_DOMAIN_RUNTIME_API: domain_runtime_callback(callback_id, cb_data); break;
+    default: break;
+  }
+}
+
+// Invoked by CUPTI when a new buffer is needed to record CUPTI activity events.
+void CUPTIAPI buffer_requested_callback(uint8_t** buffer_ptr_ptr,
+                                        size_t* size_ptr,
+                                        size_t* max_num_records_ptr)
+{
+  *max_num_records_ptr = 0;
+  if (!State->is_shutdown) {
+    auto buffer = State->free_buffers.get();
+    buffer->release(buffer_ptr_ptr, size_ptr);
+  } else {
+    *buffer_ptr_ptr = nullptr;
+    *size_ptr       = 0;
+  }
+}
+
+// Invoked by CUPTI when an activity event buffer has completed.
+void CUPTIAPI buffer_completed_callback(
+  CUcontext, uint32_t, uint8_t* buffer, size_t buffer_size, size_t valid_size)
+{
+  auto pb = std::make_unique<profile_buffer>(buffer, buffer_size, valid_size);
+  if (!State->is_shutdown) { State->completed_buffers.put(std::move(pb)); }
+}
+
+// Setup the environment variables for NVTX library injection so we can capture NVTX events.
+void setup_nvtx_env(JNIEnv* env, jstring j_lib_path)
+{
+  auto lib_path = env->GetStringUTFChars(j_lib_path, 0);
+  if (lib_path == NULL) { throw std::runtime_error("Error getting library path"); }
+  setenv("NVTX_INJECTION64_PATH", lib_path, 1);
+  env->ReleaseStringUTFChars(j_lib_path, lib_path);
+}
+
+// Main processing loop for the background writer thread
+void writer_thread_process(JavaVM* vm, jobject j_writer, size_t buffer_size, size_t flush_threshold)
+{
+  try {
+    JNIEnv* env = attach_to_jvm(vm);
+    profiler_serializer serializer(env, j_writer, buffer_size, flush_threshold);
+    auto buffer = State->completed_buffers.get();
+    while (buffer) {
+      serializer.process_cupti_buffer(buffer->data(), buffer->valid_size());
+      State->free_buffers.put(std::move(buffer));
+      buffer = State->completed_buffers.get();
+    }
+    serializer.flush();
+  } catch (std::exception const& e) {
+    std::cerr << "PROFILER: WRITER THREAD ERROR: " << e.what() << std::endl;
+    // no-op process buffers
+    auto buffer = State->completed_buffers.get();
+    while (buffer) {
+      State->free_buffers.put(std::move(buffer));
+      buffer = State->completed_buffers.get();
+    }
+  }
+  vm->DetachCurrentThread();
+}
+
+// Enable/disable capture of CUPTI activity events
+void update_activity_enable(bool enable)
+{
+  if (enable) {
+    check_cupti(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_DEVICE), "Error enabling device activity");
+    // check_cupti(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_CONTEXT), "Error enabling context
```

Comment:
Could you remove the commented out code, or add a comment explaining what is happening here.